### PR TITLE
Improve backend query export names

### PR DESCRIPTION
**Pull request by melton-jason**
> Yes, we should improve the back-end file name. But, Caroline created a separate issue for that, so it doesn't necessarily have to be fixed as part of this PR - https://github.com/specify/specify7/issues/3525

Addresses https://github.com/specify/specify7/pull/3476#issuecomment-1560024885 and part of #3525 

Follows the more user-friendly Query Export file name introduced in #3476. 
All query export files should now be in the following format: `QueryName - Weekday (abbreviated) Month (abbreviated) Day Year`
Example: `New Query - Wed Jun 07 2023`

To Test:
Without selecting any rows:
- Export a Query to csv
  - Ensure the file name has the correct format (which should be similar to the file name when selected rows are exported)
- Export a Query to kml 
  - Ensure the file name has the correct format
  - The `Export to KML` button is only visible when the Query contains Coordinate data (like Longitude1 and Latitude1)
  - No testing needs to be done regarding exporting selected rows to kml: this is not currently supported (#3589)

### CODE REVIEW

**Review by carlosmbe — CHANGES_REQUESTED** [review]
Two Things:
Exported Files are not CSV when using `All Items` Create CSV. It appears they have no extension and are treated as a "Document" on Mac. This happens to both Saved Queries and New Queries.

Secondly, when it's a New Query which is not saved. The Type is included with `Selected` and not with the `All Items` option.

Here is an example made from the same query using both options

- Selected Items `New Query Locality - Mon Jun 12 2023.csv`

- All Items `New Query - Mon Jun 12 2023`

**As for the KML Exports**

The Export to KML button is only visible when the Query contains Coordinate data. It would be great if we could resolve #3518
Need to test how Google Earth is accepting Exports

**Comment by melton-jason** [comment]
> Two Things: Exported Files are not CSV when using `All Items` Create CSV. It appears they have no extension and are treated as a "Document" on Mac. This happens to both Saved Queries and New Queries.
> 
> Secondly, when it's a New Query which is not saved. The Type is included with `Selected` and not with the `All Items` option.
> 
> Here is an example made from the same query using both options
> 
> * Selected Items `New Query Locality - Mon Jun 12 2023.csv`
> * All Items `New Query - Mon Jun 12 2023`
> 
> **As for the KML Exports**
> 
> The Export to KML button is only visible when the Query contains Coordinate data. It would be great if we could resolve #3518 Need to test how Google Earth is accepting Exports

Both of these issues should be fixed. Can this be tested again? 

To be consistent, I also hid the Export to CSV button when there are no Query Results. 
If we still want the buttons visible on the Query Results, we can disable them instead of hiding them. 
@specify/ux-testing, let me know which option is preferable.

**Review by carlosmbe — CHANGES_REQUESTED** [review]
Nice work fixing the lack of extension part. However this is still present:

When it's a New Query which is not saved. The Type is included with `Selected` and not with the `All Items` export option.

**Review by emenslin — APPROVED** [review]
Both csv and kml names have the correct format, looks good and this seems like it'll make things a lot easier for the users. 
Screenshot 2024-01-11 155102

File: specifyweb/frontend/js_src/lib/components/QueryBuilder/Export.tsx [hunk 1]
```diff
@@ -100,9 +100,14 @@ export function QueryExportButtons({
     return resultToExport.join('\n');
   }
 
+  const containsResults =
+    (results.current?.length ?? 0) > 0 &&
```
Comment (maxpatiiuk) [inline specifyweb/frontend/js_src/lib/components/QueryBuilder/Export.tsx:104]: hmm, this line seems redundant with the next one
(.every() returns false for empty arrays)

Comment (melton-jason) [reply specifyweb/frontend/js_src/lib/components/QueryBuilder/Export.tsx:109]: The type of results is `React.MutableRefObject<RA<QueryResultRow | undefined> | undefined>`, thus the line `results.current?.every((row) => row !== undefined)` handles the case when `results.current` is an array and contains elements, but may have `undefined` elements.

Comment (maxpatiiuk) [reply specifyweb/frontend/js_src/lib/components/QueryBuilder/Export.tsx:109]: ups, I was wrong about `(.every() returns false for empty arrays)` 🤦

.every() returns TRUE for empty arrays

it's `.some()` that returns false for empty

---

thus, in this case, could you use `results.current?.some(row => row !== undefined)` instead?

Comment (melton-jason) [reply specifyweb/frontend/js_src/lib/components/QueryBuilder/Export.tsx]: Fixed in https://github.com/specify/specify7/pull/3590/commits/0a88e94fb07f3c92adf821397c97dca57f14d34e

**Review by maxpatiiuk — APPROVED** [review]

**Review by grantfitzsimmons — APPROVED** [review]
Fun note (not that I think it's worth holding up this fix for)– exports done on the front-end show Wednesday as the day, exports done on the back-end show Thursday as the day.

image

Screenshot 2024-01-17 at 9 02 32 PM

Screenshot 2024-01-17 at 9 03 21 PM

**Comment by grantfitzsimmons** [comment]
> When it's a New Query which is not saved. The Type is included with `Selected` and not with the `All Items` export option.

@melton-jason Is this something that you have resolved?

**Review by carlosmbe — COMMENTED** [review]
Not a bug (I think) but when I did an `All Export`, the day was Thursday 18 Jan. However, the selected export was Wednesday 17 Jan. 

I did this test at 11PM on Wednesday

The all export one is the 2MB file with 18 Jan as the file name 

Screenshot 2024-01-17 at 11 18 08 PM

**Comment by melton-jason** [comment]
> > When it's a New Query which is not saved. The Type is included with `Selected` and not with the `All Items` export option.
> 
> @melton-jason Is this something that you have resolved?

This is now resolved. 



> Not a bug (I think) but when I did an `All Export`, the day was Thursday 18 Jan. However, the selected export was Wednesday 17 Jan.
> 
> I did this test at 11PM on Wednesday
> 
> The all export one is the 2MB file with 18 Jan as the file name
> 
> Screenshot 2024-01-17 at 11 18 08 PM

This is very likely caused by #641

**Review by emenslin — APPROVED** [review]
This isn't a big enough problem where I'm going to request changes however when you select rows (bottom) vs not select rows (top) one has a space between 'collection object' and the other doesn't 
Screenshot 2024-01-25 123055

**Comment by melton-jason** [comment]
This discrepancy is due to the fact that the frontend uses the localized label for the table and the backend uses the `contextname` of the Query. 
The backend currently has no way to easily get the table localization without re-running the SQL to fetch it. 
(The localization used to be cached, but this was intentionally removed in https://github.com/specify/specify7/commit/b2c982506314cf3d75ff1f6881e0f42413d9ae88). 

At the moment, it may be unwise to re-implement a caching mechanism on the backend. 

So the easiest way to resolve this discrepancy would be to use the database table name on the frontend, rather than the localized table name. 
@specify/ux-testing, would this change be worth maintaining consistency for export file names, should it be kept the same, or should an implementation be made on the backend to easily fetch the localization without hitting the database when not needed?

**Review by carlosmbe — APPROVED** [review]

**Pull request merged**